src/Stateless/StateConfiguration.cs

```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
             /// <returns>The receiver.</returns>
             public StateConfiguration SubstateOf(TState superstate)
             {
+                var State = _representation.UnderlyingState;
+
+                // Check for accidental identical cyclic configuration 
+                if (State.Equals(superstate)) { throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal cyclic configuration."); }
+
+                // Check for accidental identical nested cyclic configuration 
+                List<TState> superstates = new List<TState>();
```

Comment:
Thinking through it a little further, could be:

```csharp
var superstates = new HashSet<TState>{ State };
```

The check for membership could then be moved inside the `while` loop as below.